src/openrct2/core/Http.WinHttp.cpp

```diff
@@ -0,0 +1,223 @@
+
+            DWORD dwDownloaded{};
+            if (!WinHttpReadData(hRequest, dst, dwSize, &dwDownloaded))
+                ThrowWin32Exception("WinHttpReadData");
+
+            // This condition should never be reached since WinHttpQueryDataAvailable
+            // reported that there are bits to read.
+            if (dwDownloaded == 0)
+                break;
+        } while (dwSize > 0);
+        return body;
+    }
+
+    Response Do(const Request& req)
+    {
+        HINTERNET hSession{}, hConnect{}, hRequest{};
+        try
+        {
+            URL_COMPONENTS url;
+            ZeroMemory(&url, sizeof(url));
+            url.dwStructSize = sizeof(url);
+            url.dwSchemeLength = (DWORD)-1;
+            url.dwHostNameLength = (DWORD)-1;
+            url.dwUrlPathLength = (DWORD)-1;
+            url.dwExtraInfoLength = (DWORD)-1;
+
+            auto wUrl = String::ToWideChar(req.url);
+            if (!WinHttpCrackUrl(wUrl.c_str(), 0, 0, &url))
+                throw std::invalid_argument("Unable to parse URI.");
+
+            auto userAgent = String::ToWideChar(OPENRCT2_USER_AGENT);
+            hSession = WinHttpOpen(
+                userAgent.c_str(), WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
+            if (hSession == nullptr)
+                ThrowWin32Exception("WinHttpOpen");
+
+            auto wHostName = std::wstring(url.lpszHostName, url.dwHostNameLength);
+            hConnect = WinHttpConnect(hSession, wHostName.c_str(), url.nPort, 0);
+            if (hConnect == nullptr)
+                ThrowWin32Exception("WinHttpConnect");
+
+            auto wVerb = GetVerb(req.method);
+            auto wQuery = std::wstring(url.lpszUrlPath, url.dwUrlPathLength);
+            hRequest = WinHttpOpenRequest(
+                hConnect, wVerb, wQuery.c_str(), NULL, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, WINHTTP_FLAG_SECURE);
+            if (hRequest == nullptr)
+                ThrowWin32Exception("WinHttpOpenRequest");
+
+            for (auto header : req.header)
+            {
+                auto fullHeader = String::ToWideChar(header.first) + L": " + String::ToWideChar(header.second);
+                if (!WinHttpAddRequestHeaders(hRequest, fullHeader.c_str(), (ULONG)-1L, WINHTTP_ADDREQ_FLAG_ADD))
+                    ThrowWin32Exception("WinHttpAddRequestHeaders");
+            }
+
+            auto reqBody = (LPVOID)req.body.data();
+            auto reqBodyLen = (DWORD)req.body.size();
+            if (!WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0, reqBody, reqBodyLen, reqBodyLen, 0))
+                ThrowWin32Exception("WinHttpSendRequest");
+
+            if (!WinHttpReceiveResponse(hRequest, NULL))
+                ThrowWin32Exception("WinHttpReceiveResponse");
+
+            auto statusCode = ReadStatusCode(hRequest);
+            auto contentType = ReadHeader(hRequest, WINHTTP_QUERY_CONTENT_TYPE, L"Content-Type");
```

Comment:
Could probably save some code by moving this line after ReadHeaders and then use the map to get this field.